// @c15t/logger/dist/__tests__/core/types.test.cjs
"use strict";
var __webpack_exports__ = {};
const external_vitest_namespaceObject = require("vitest");
(0, external_vitest_namespaceObject.describe)('types', ()=>{
    (0, external_vitest_namespaceObject.describe)('LogLevel', ()=>{
        (0, external_vitest_namespaceObject.it)('should allow valid log levels', ()=>{
            const levels = [
                'info',
                'success',
                'warn',
                'error',
                'debug'
            ];
            (0, external_vitest_namespaceObject.expect)(levels).toHaveLength(5);
        });
    });
    (0, external_vitest_namespaceObject.describe)('LoggerOptions', ()=>{
        (0, external_vitest_namespaceObject.it)('should allow creating valid logger options', ()=>{
            const options = {
                disabled: false,
                level: 'warn',
                log: (level, message, ...args)=>{
                    console.log(level, message, args);
                }
            };
            (0, external_vitest_namespaceObject.expect)(options).toHaveProperty('disabled');
            (0, external_vitest_namespaceObject.expect)(options).toHaveProperty('level');
            (0, external_vitest_namespaceObject.expect)(options).toHaveProperty('log');
            const emptyOptions = {};
            (0, external_vitest_namespaceObject.expect)(emptyOptions).not.toBeNull();
        });
    });
    (0, external_vitest_namespaceObject.describe)('Logger', ()=>{
        (0, external_vitest_namespaceObject.it)('should have methods for all log levels', ()=>{
            const logger = {
                info: (_message, ..._args)=>{},
                success: (_message, ..._args)=>{},
                warn: (_message, ..._args)=>{},
                error: (_message, ..._args)=>{},
                debug: (_message, ..._args)=>{}
            };
            (0, external_vitest_namespaceObject.expect)(logger).toHaveProperty('info');
            (0, external_vitest_namespaceObject.expect)(logger).toHaveProperty('success');
            (0, external_vitest_namespaceObject.expect)(logger).toHaveProperty('warn');
            (0, external_vitest_namespaceObject.expect)(logger).toHaveProperty('error');
            (0, external_vitest_namespaceObject.expect)(logger).toHaveProperty('debug');
        });
    });
    (0, external_vitest_namespaceObject.describe)('LoggableError', ()=>{
        (0, external_vitest_namespaceObject.it)('should allow creating error objects with required properties', ()=>{
            const minimalError = {
                message: 'Test error'
            };
            (0, external_vitest_namespaceObject.expect)(minimalError).toHaveProperty('message');
            const fullError = {
                message: 'Test error',
                code: 'TEST_ERROR',
                status: 400,
                meta: {
                    test: true
                },
                category: 'test',
                stack: 'Error stack trace'
            };
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('message');
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('code');
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('status');
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('meta');
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('category');
            (0, external_vitest_namespaceObject.expect)(fullError).toHaveProperty('stack');
        });
    });
});
for(var __webpack_i__ in __webpack_exports__)exports[__webpack_i__] = __webpack_exports__[__webpack_i__];
Object.defineProperty(exports, '__esModule', {
    value: true
});
