// @orpc/openapi/dist/index.mjs
import { c as customOpenAPIOperation } from './shared/openapi.1iT1iSZi.mjs';
export { C as CompositeSchemaConverter, L as LOGIC_KEYWORDS, O as OpenAPIGenerator, a as applyCustomOpenAPIOperation, n as applySchemaOptionality, h as checkParamsSchema, p as expandArrayableSchema, o as expandUnionSchema, m as filterSchemaBranches, g as getCustomOpenAPIOperation, l as isAnySchema, j as isFileSchema, k as isObjectSchema, q as isPrimitiveSchema, r as resolveOpenAPIJsonSchemaRef, s as separateObjectSchema, d as toOpenAPIContent, e as toOpenAPIEventIteratorContent, b as toOpenAPIMethod, f as toOpenAPIParameters, t as toOpenAPIPath, i as toOpenAPISchema } from './shared/openapi.1iT1iSZi.mjs';
import { createORPCErrorFromJson } from '@orpc/client';
import { StandardOpenAPISerializer, StandardOpenAPIJsonSerializer, StandardBracketNotationSerializer } from '@orpc/openapi-client/standard';
import { ORPCError, createRouterClient } from '@orpc/server';
import { resolveMaybeOptionalOptions } from '@orpc/shared';
export { ContentEncoding as JSONSchemaContentEncoding, Format as JSONSchemaFormat, TypeName as JSONSchemaTypeName } from '@orpc/interop/json-schema-typed/draft-2020-12';
import '@orpc/client/standard';
import '@orpc/contract';

function createJsonifiedRouterClient(router, ...rest) {
  const options = resolveMaybeOptionalOptions(rest);
  const serializer = new StandardOpenAPISerializer(new StandardOpenAPIJsonSerializer(), new StandardBracketNotationSerializer());
  options.interceptors ??= [];
  options.interceptors.unshift(async (options2) => {
    try {
      return serializer.deserialize(
        serializer.serialize(
          await options2.next()
        )
      );
    } catch (e) {
      if (e instanceof ORPCError) {
        throw createORPCErrorFromJson(serializer.deserialize(
          serializer.serialize(
            e.toJSON(),
            { outputFormat: "plain" }
          )
        ));
      }
      throw e;
    }
  });
  return createRouterClient(router, options);
}

const oo = {
  spec: customOpenAPIOperation
};

export { createJsonifiedRouterClient, customOpenAPIOperation, oo };
