// fumadb/dist/index.js
import {
  createNameVariantsBuilder
} from "./chunk-IVWNQO5P.js";

// src/index.ts
import { compare } from "semver";

// src/shared/providers.ts
var sqlProviders = [
  "sqlite",
  "cockroachdb",
  "mysql",
  "postgresql",
  "mssql"
];
var noSqlProviders = ["mongodb"];
var providers = [...sqlProviders, ...noSqlProviders];

// src/index.ts
function fumadb(config) {
  const schemas = config.schemas.sort((a, b) => compare(a.version, b.version));
  return {
    names: createNameVariantsBuilder(config.namespace, schemas, (schemas2) => {
      return fumadb({
        ...config,
        schemas: schemas2
      });
    }),
    version(targetVersion) {
      return targetVersion;
    },
    client(adapter) {
      const orms = /* @__PURE__ */ new Map();
      const adapterContext = {
        ...config
      };
      return {
        adapter,
        schemas,
        orm(version) {
          let orm = orms.get(version);
          if (orm) return orm;
          const schema = schemas.find((schema2) => schema2.version === version);
          if (!schema) throw new Error(`unknown schema version ${version}`);
          orm = adapter.createORM.call(adapterContext, schema);
          orms.set(version, orm);
          return orm;
        },
        async version() {
          const version = await adapter.getSchemaVersion.call(adapterContext);
          if (!version)
            throw new Error(`FumaDB ${config.namespace} is not initialized.`);
          return version;
        },
        generateSchema(version, name = config.namespace) {
          if (!adapter.generateSchema)
            throw new Error("The adapter doesn't support schema API.");
          let schema;
          if (version === "latest") {
            schema = schemas.at(-1);
          } else {
            schema = schemas.find((schema2) => schema2.version === version);
            if (!schema) throw new Error(`Invalid version: ${version}`);
          }
          return adapter.generateSchema.call(adapterContext, schema, name);
        },
        createMigrator() {
          if (!adapter.createMigrationEngine)
            throw new Error("The adapter doesn't support migration engine.");
          return adapter.createMigrationEngine.call(adapterContext);
        },
        get abstract() {
          return this.orm(schemas.at(-1).version);
        }
      };
    }
  };
}
export {
  fumadb,
  noSqlProviders,
  providers,
  sqlProviders
};
