// fumadocs-core/dist/framework/react-router.js
import {
  FrameworkProvider
} from "../chunk-K4WNLOVQ.js";
import "../chunk-U67V476Y.js";

// src/framework/react-router.tsx
import { useMemo } from "react";
import {
  Link,
  useLocation,
  useNavigate,
  useParams,
  useRevalidator
} from "react-router";
import { jsx } from "react/jsx-runtime";
var framework = {
  usePathname() {
    return useLocation().pathname;
  },
  useParams() {
    return useParams();
  },
  useRouter() {
    const navigate = useNavigate();
    const revalidator = useRevalidator();
    return useMemo(
      () => ({
        push(url) {
          navigate(url);
        },
        refresh() {
          void revalidator.revalidate();
        }
      }),
      [navigate, revalidator]
    );
  },
  Link({ href, prefetch, ...props }) {
    return /* @__PURE__ */ jsx(Link, { to: href, prefetch: prefetch ? "intent" : "none", ...props, children: props.children });
  }
};
function ReactRouterProvider({
  children,
  Link: CustomLink,
  Image: CustomImage
}) {
  return /* @__PURE__ */ jsx(
    FrameworkProvider,
    {
      ...framework,
      Link: CustomLink ?? framework.Link,
      Image: CustomImage ?? framework.Image,
      children
    }
  );
}
export {
  ReactRouterProvider
};
