// fumadocs-core/dist/framework/tanstack.js
import {
  FrameworkProvider
} from "../chunk-K4WNLOVQ.js";
import "../chunk-U67V476Y.js";

// src/framework/tanstack.tsx
import { useRef, useMemo } from "react";
import {
  useParams,
  Link,
  useRouter,
  useRouterState
} from "@tanstack/react-router";
import { jsx } from "react/jsx-runtime";
var framework = {
  Link({ href, prefetch = true, ...props }) {
    return /* @__PURE__ */ jsx(Link, { to: href, preload: prefetch ? "intent" : false, ...props, children: props.children });
  },
  usePathname() {
    const { isLoading, pathname } = useRouterState({
      select: (state) => ({
        isLoading: state.isLoading,
        pathname: state.location.pathname
      })
    });
    const activePathname = useRef(pathname);
    return useMemo(() => {
      if (isLoading) {
        return activePathname.current;
      }
      activePathname.current = pathname;
      return pathname;
    }, [isLoading, pathname]);
  },
  useRouter() {
    const router = useRouter();
    return useMemo(
      () => ({
        push(url) {
          void router.navigate({
            href: url
          });
        },
        refresh() {
          void router.invalidate();
        }
      }),
      [router]
    );
  },
  useParams() {
    return useParams({ strict: false });
  }
};
function TanstackProvider({
  children,
  Link: CustomLink,
  Image: CustomImage
}) {
  return /* @__PURE__ */ jsx(
    FrameworkProvider,
    {
      ...framework,
      Link: CustomLink ?? framework.Link,
      Image: CustomImage ?? framework.Image,
      children
    }
  );
}
export {
  TanstackProvider
};
