// @c15t/backend/src/v2/db/schema/1.0.0/subject.ts
import { column, idColumn, table } from 'fumadb/schema';
import { z } from 'zod';

export const subjectTable = table('subject', {
	id: idColumn('id', 'varchar(255)'),
	isIdentified: column('isIdentified', 'bool').defaultTo$(() => false),
	externalId: column('externalId', 'string').nullable(),
	identityProvider: column('identityProvider', 'string').nullable(),
	lastIpAddress: column('lastIpAddress', 'string').nullable(),
	subjectTimezone: column('subjectTimezone', 'string').nullable(),
	createdAt: column('createdAt', 'timestamp').defaultTo$('now'),
	updatedAt: column('updatedAt', 'timestamp').defaultTo$('now'),
});

export const subjectSchema = z.object({
	id: z.string(),
	isIdentified: z.boolean().prefault(false),
	externalId: z.string().nullish(),
	identityProvider: z.string().nullish(),
	lastIpAddress: z.string().optional(),
	subjectTimezone: z.string().nullish(),
	createdAt: z.date().prefault(() => new Date()),
	updatedAt: z.date().prefault(() => new Date()),
});

export type Subject = z.infer<typeof subjectSchema>;
