// @bprogress/next/dist/app.cjs
"use strict";
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
function _object_without_properties(source, excluded) {
    if (source == null) return {};
    var target = _object_without_properties_loose(source, excluded);
    var key, i;
    if (Object.getOwnPropertySymbols) {
        var sourceSymbolKeys = Object.getOwnPropertySymbols(source);
        for(i = 0; i < sourceSymbolKeys.length; i++){
            key = sourceSymbolKeys[i];
            if (excluded.indexOf(key) >= 0) continue;
            if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
            target[key] = source[key];
        }
    }
    return target;
}
function _object_without_properties_loose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;
    for(i = 0; i < sourceKeys.length; i++){
        key = sourceKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        target[key] = source[key];
    }
    return target;
}
function _type_of(obj) {
    "@swc/helpers - typeof";
    return obj && typeof Symbol !== "undefined" && obj.constructor === Symbol ? "symbol" : typeof obj;
}
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = function(target, all) {
    for(var name in all)__defProp(target, name, {
        get: all[name],
        enumerable: true
    });
};
var __copyProps = function(to, from, except, desc) {
    if (from && (typeof from === "undefined" ? "undefined" : _type_of(from)) === "object" || typeof from === "function") {
        var _iteratorNormalCompletion = true, _didIteratorError = false, _iteratorError = undefined;
        try {
            var _loop = function() {
                var key = _step.value;
                if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
                    get: function() {
                        return from[key];
                    },
                    enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
                });
            };
            for(var _iterator = __getOwnPropNames(from)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true)_loop();
        } catch (err) {
            _didIteratorError = true;
            _iteratorError = err;
        } finally{
            try {
                if (!_iteratorNormalCompletion && _iterator.return != null) {
                    _iterator.return();
                }
            } finally{
                if (_didIteratorError) {
                    throw _iteratorError;
                }
            }
        }
    }
    return to;
};
var __toESM = function(mod, isNodeMode, target) {
    return target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(// If the importer is in node compatibility mode or this is not an ESM
    // file that has been converted to a CommonJS file using a Babel-
    // compatible transform (i.e. "__esModule" has not been set), then set
    // "default" to the CommonJS "module.exports" for node compatibility.
    isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
        value: mod,
        enumerable: true
    }) : target, mod);
};
var __toCommonJS = function(mod) {
    return __copyProps(__defProp({}, "__esModule", {
        value: true
    }), mod);
};
// src/app.ts
var app_exports = {};
__export(app_exports, {
    ProgressProvider: function() {
        return AppProgressProvider;
    },
    useRouter: function() {
        return useRouter;
    }
});
module.exports = __toCommonJS(app_exports);
// src/hooks/use-router.ts
var import_navigation = require("next/navigation.js");
var import_react = require("react");
var import_core = require("@bprogress/core");
var import_react2 = require("@bprogress/react");
function removeFirstPathSegment(url) {
    var parts = url.pathname.split("/");
    if (parts.length > 1 && parts[1]) {
        parts.splice(1, 1);
        url.pathname = parts.join("/") || "/";
    }
    return url;
}
function useRouter(options) {
    var _ref = options || {}, customRouter = _ref.customRouter, defaultProgressOptions = _object_without_properties(_ref, [
        "customRouter"
    ]);
    var router = customRouter ? customRouter() : // eslint-disable-next-line react-hooks/rules-of-hooks
    (0, import_navigation.useRouter)();
    var _ref1 = (0, import_react2.useProgress)(), start = _ref1.start, stop = _ref1.stop, providerDisableSameURL = _ref1.disableSameURL, providerStartPosition = _ref1.startPosition, providerDelay = _ref1.delay, providerStopDelay = _ref1.stopDelay;
    var extendedRouterRef = (0, import_react.useRef)(null);
    function createHandler(fn) {
        return function(href, options2) {
            var _ref = options2 || {}, showProgress = _ref.showProgress, startPosition = _ref.startPosition, disableSameURL = _ref.disableSameURL, basePath = _ref.basePath, i18nPath = _ref.i18nPath, delay = _ref.delay, stopDelay = _ref.stopDelay, routerOpts = _object_without_properties(_ref, [
                "showProgress",
                "startPosition",
                "disableSameURL",
                "basePath",
                "i18nPath",
                "delay",
                "stopDelay"
            ]);
            var progressOpts = _object_spread_props(_object_spread({}, defaultProgressOptions), {
                showProgress: showProgress,
                startPosition: startPosition,
                disableSameURL: disableSameURL,
                basePath: basePath,
                i18nPath: i18nPath,
                delay: delay,
                stopDelay: stopDelay
            });
            var localDisableSameURL = progressOpts.disableSameURL !== void 0 ? progressOpts.disableSameURL : providerDisableSameURL;
            var localStartPosition = progressOpts.startPosition !== void 0 ? progressOpts.startPosition : providerStartPosition;
            var localDelay = progressOpts.delay !== void 0 ? progressOpts.delay : providerDelay;
            var localStopDelay = progressOpts.stopDelay !== void 0 ? progressOpts.stopDelay : providerStopDelay;
            if (progressOpts.showProgress === false) {
                return fn(href, routerOpts);
            }
            var currentUrl = new URL(location.href);
            var targetUrl = new URL(href, location.href);
            if (progressOpts.i18nPath) {
                currentUrl = removeFirstPathSegment(currentUrl);
            }
            if (progressOpts.basePath) {
                targetUrl.pathname = progressOpts.basePath + (targetUrl.pathname !== "/" ? targetUrl.pathname : "");
            }
            var sameURL = (0, import_core.isSameURL)(targetUrl, currentUrl);
            if (sameURL && localDisableSameURL) {
                return fn(href, routerOpts);
            }
            start(localStartPosition, localDelay);
            setTimeout(function() {
                if (sameURL) stop(localStopDelay);
            }, localDelay || 0);
            return fn(href, routerOpts);
        };
    }
    function createNoHrefHandler(fn) {
        return function(options2) {
            var _ref = options2 || {}, showProgress = _ref.showProgress, startPosition = _ref.startPosition, disableSameURL = _ref.disableSameURL, basePath = _ref.basePath, i18nPath = _ref.i18nPath, delay = _ref.delay, stopDelay = _ref.stopDelay, routerOpts = _object_without_properties(_ref, [
                "showProgress",
                "startPosition",
                "disableSameURL",
                "basePath",
                "i18nPath",
                "delay",
                "stopDelay"
            ]);
            var progressOpts = _object_spread_props(_object_spread({}, defaultProgressOptions), {
                showProgress: showProgress,
                startPosition: startPosition,
                disableSameURL: disableSameURL,
                basePath: basePath,
                i18nPath: i18nPath,
                delay: delay,
                stopDelay: stopDelay
            });
            var localStartPosition = progressOpts.startPosition !== void 0 ? progressOpts.startPosition : providerStartPosition;
            var localDelay = progressOpts.delay !== void 0 ? progressOpts.delay : providerDelay;
            var localStopDelay = progressOpts.stopDelay !== void 0 ? progressOpts.stopDelay : providerStopDelay;
            if (progressOpts.showProgress === false) {
                return fn(routerOpts);
            }
            start(localStartPosition, localDelay);
            var result = fn(routerOpts);
            setTimeout(function() {
                stop(localStopDelay);
            }, localDelay || 0);
            return result;
        };
    }
    function createPrefetchHandler(fn) {
        return function(href, options2) {
            return fn(href, options2);
        };
    }
    if (!extendedRouterRef.current) {
        extendedRouterRef.current = _object_spread_props(_object_spread({}, router), {
            push: createHandler(router.push),
            replace: createHandler(router.replace),
            prefetch: createPrefetchHandler(router.prefetch),
            back: createNoHrefHandler(router.back),
            refresh: createNoHrefHandler(router.refresh),
            forward: createNoHrefHandler(router.forward)
        });
    } else {
        extendedRouterRef.current.push = createHandler(router.push);
        extendedRouterRef.current.replace = createHandler(router.replace);
        extendedRouterRef.current.prefetch = createPrefetchHandler(router.prefetch);
        extendedRouterRef.current.back = createNoHrefHandler(router.back);
        extendedRouterRef.current.refresh = createNoHrefHandler(router.refresh);
        extendedRouterRef.current.forward = createNoHrefHandler(router.forward);
    }
    return extendedRouterRef.current;
}
// src/providers/app-progress-provider.tsx
var import_react6 = __toESM(require("react"), 1);
// src/components/app-progress.tsx
var import_react3 = require("@bprogress/react");
var import_navigation2 = require("next/navigation.js");
var AppProgressComponent = function(props) {
    var pathname = (0, import_navigation2.usePathname)();
    var searchParams = (0, import_navigation2.useSearchParams)();
    (0, import_react3.useAnchorProgress)(props, [
        pathname,
        searchParams
    ]);
    return null;
};
var AppProgress = (0, import_react3.withMemo)(AppProgressComponent);
AppProgress.displayName = "AppProgress";
// src/providers/app-progress-provider.tsx
var import_react7 = require("@bprogress/react");
// src/providers/next-progress-provider.tsx
var import_react4 = __toESM(require("react"), 1);
var import_react5 = require("@bprogress/react");
var NextProgressProvider = function(_param) {
    var children = _param.children, ProgressComponent = _param.ProgressComponent, color = _param.color, height = _param.height, options = _param.options, spinnerPosition = _param.spinnerPosition, style = _param.style, disableStyle = _param.disableStyle, nonce = _param.nonce, stopDelay = _param.stopDelay, delay = _param.delay, startPosition = _param.startPosition, disableSameURL = _param.disableSameURL, shallowRouting = _param.shallowRouting, props = _object_without_properties(_param, [
        "children",
        "ProgressComponent",
        "color",
        "height",
        "options",
        "spinnerPosition",
        "style",
        "disableStyle",
        "nonce",
        "stopDelay",
        "delay",
        "startPosition",
        "disableSameURL",
        "shallowRouting"
    ]);
    return /* @__PURE__ */ import_react4.default.createElement(import_react5.ProgressProvider, {
        color: color,
        height: height,
        options: options,
        spinnerPosition: spinnerPosition,
        style: style,
        disableStyle: disableStyle,
        nonce: nonce,
        stopDelay: stopDelay,
        delay: delay,
        startPosition: startPosition,
        disableSameURL: disableSameURL,
        shallowRouting: shallowRouting
    }, /* @__PURE__ */ import_react4.default.createElement(ProgressComponent, _object_spread({
        stopDelay: stopDelay,
        delay: delay,
        startPosition: startPosition,
        disableSameURL: disableSameURL,
        shallowRouting: shallowRouting
    }, props)), children);
};
// src/providers/app-progress-provider.tsx
var SuspendedAppProgress = (0, import_react7.withSuspense)(AppProgress);
var AppProgressProvider = function(_param) {
    var children = _param.children, props = _object_without_properties(_param, [
        "children"
    ]);
    return /* @__PURE__ */ import_react6.default.createElement(NextProgressProvider, _object_spread({
        ProgressComponent: SuspendedAppProgress
    }, props), children);
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
    ProgressProvider: ProgressProvider,
    useRouter: useRouter
});
