// fumadocs-core/dist/chunk-K4WNLOVQ.js
// src/framework/index.tsx
import { createContext, use, useMemo } from "react";
import { jsx } from "react/jsx-runtime";
var notImplemented = () => {
  throw new Error(
    "You need to wrap your application inside `FrameworkProvider`."
  );
};
var FrameworkContext = createContext({
  useParams: notImplemented,
  useRouter: notImplemented,
  usePathname: notImplemented
});
function FrameworkProvider({
  Link: Link2,
  useRouter: useRouter2,
  useParams: useParams2,
  usePathname: usePathname2,
  Image: Image2,
  children
}) {
  const framework = useMemo(
    () => ({
      usePathname: usePathname2,
      useRouter: useRouter2,
      Link: Link2,
      Image: Image2,
      useParams: useParams2
    }),
    [Link2, usePathname2, useRouter2, useParams2, Image2]
  );
  return /* @__PURE__ */ jsx(FrameworkContext, { value: framework, children });
}
function usePathname() {
  return use(FrameworkContext).usePathname();
}
function useRouter() {
  return use(FrameworkContext).useRouter();
}
function useParams() {
  return use(FrameworkContext).useParams();
}
function Image(props) {
  const { Image: Image2 } = use(FrameworkContext);
  if (!Image2) {
    const { src, alt, priority, ...rest } = props;
    return /* @__PURE__ */ jsx(
      "img",
      {
        alt,
        src,
        fetchPriority: priority ? "high" : "auto",
        ...rest
      }
    );
  }
  return /* @__PURE__ */ jsx(Image2, { ...props });
}
function Link(props) {
  const { Link: Link2 } = use(FrameworkContext);
  if (!Link2) {
    const { href, prefetch: _, ...rest } = props;
    return /* @__PURE__ */ jsx("a", { href, ...rest });
  }
  return /* @__PURE__ */ jsx(Link2, { ...props });
}

export {
  FrameworkProvider,
  usePathname,
  useRouter,
  useParams,
  Image,
  Link
};
