// @c15t/backend/dist/pkgs/db-adapters/adapters/kysely-adapter/index.js
import { Kysely, MssqlDialect, MysqlDialect, PostgresDialect, SqliteDialect } from "kysely";
import superjson from "superjson";
import base_x from "base-x";
import { z } from "zod";
import "neverthrow";
import "@orpc/server";
import { trace } from "@opentelemetry/api";
import "@c15t/logger";
import "drizzle-orm";
function getDatabaseType(db) {
    if (!db) return null;
    if ('dialect' in db) return getDatabaseType(db.dialect);
    if (db && 'object' == typeof db && 'createDriver' in db) {
        if (db instanceof SqliteDialect) return 'sqlite';
        if (db instanceof MysqlDialect) return 'mysql';
        if (db instanceof PostgresDialect) return 'postgres';
        if (db instanceof MssqlDialect) return 'mssql';
    }
    if (db && 'object' == typeof db && 'aggregate' in db) return 'sqlite';
    if (db && 'object' == typeof db && 'getConnection' in db) return 'mysql';
    if (db && 'object' == typeof db && 'connect' in db) return 'postgres';
    return null;
}
const dialect_createKyselyAdapter = async (config)=>{
    const db = config.database;
    if (!db) return {
        kysely: null,
        databaseType: null
    };
    if (db && 'object' == typeof db && 'db' in db) {
        const kyselyConfig = db;
        return {
            kysely: kyselyConfig.db,
            databaseType: kyselyConfig.type
        };
    }
    if (db && 'object' == typeof db && 'dialect' in db) {
        const dialectConfig = db;
        return {
            kysely: new Kysely({
                dialect: dialectConfig.dialect
            }),
            databaseType: dialectConfig.type
        };
    }
    let dialect;
    const databaseType = getDatabaseType(db);
    if (db && 'object' == typeof db && 'createDriver' in db) dialect = db;
    if (db && 'object' == typeof db && 'aggregate' in db) dialect = new SqliteDialect({
        database: db
    });
    if (db && 'object' == typeof db && 'getConnection' in db) dialect = new MysqlDialect({
        pool: db
    });
    if (db && 'object' == typeof db && 'connect' in db) dialect = new PostgresDialect({
        pool: db
    });
    return {
        kysely: dialect ? new Kysely({
            dialect
        }) : null,
        databaseType
    };
};
const COMMON_TIMEZONES = {
    UTC: 'UTC',
    GMT: 'GMT',
    EASTERN: 'America/New_York',
    CENTRAL: 'America/Chicago',
    MOUNTAIN: 'America/Denver',
    PACIFIC: 'America/Los_Angeles',
    LONDON: 'Europe/London',
    PARIS: 'Europe/Paris',
    BERLIN: 'Europe/Berlin',
    TOKYO: 'Asia/Tokyo',
    SHANGHAI: 'Asia/Shanghai',
    SINGAPORE: 'Asia/Singapore',
    SYDNEY: 'Australia/Sydney',
    SAO_PAULO: 'America/Sao_Paulo'
};
const b58 = base_x('123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz');
function id_generator_generateId(prefix) {
    const buf = crypto.getRandomValues(new Uint8Array(20));
    const EPOCH_TIMESTAMP = 1700000000000;
    const t = Date.now() - EPOCH_TIMESTAMP;
    const high = Math.floor(t / 0x100000000);
    const low = t >>> 0;
    buf[0] = high >>> 24 & 255;
    buf[1] = high >>> 16 & 255;
    buf[2] = high >>> 8 & 255;
    buf[3] = 255 & high;
    buf[4] = low >>> 24 & 255;
    buf[5] = low >>> 16 & 255;
    buf[6] = low >>> 8 & 255;
    buf[7] = 255 & low;
    return `${prefix}_${b58.encode(buf)}`;
}
const fieldConfigSchema = z.object({
    required: z.boolean().prefault(true),
    returned: z.boolean().prefault(true),
    input: z.boolean().prefault(true),
    defaultValue: z.any().optional(),
    transform: z.object({
        input: z.any().optional(),
        output: z.any().optional()
    }).optional(),
    validator: z.any().optional(),
    unique: z.boolean().optional(),
    indexed: z.boolean().optional(),
    sortable: z.boolean().prefault(true),
    fieldName: z.string().optional(),
    bigint: z.boolean().prefault(false)
});
const stringFieldSchema = fieldConfigSchema.extend({
    type: z.literal('string'),
    minLength: z.number().optional(),
    maxLength: z.number().optional(),
    pattern: z.string().optional()
});
const numberFieldSchema = fieldConfigSchema.extend({
    type: z.literal('number'),
    min: z.number().optional(),
    max: z.number().optional()
});
const booleanFieldSchema = fieldConfigSchema.extend({
    type: z.literal('boolean')
});
const dateFieldSchema = fieldConfigSchema.extend({
    type: z.literal('date'),
    minDate: z.date().optional(),
    maxDate: z.date().optional(),
    dateOnly: z.boolean().prefault(false),
    format: z.record(z.string(), z.unknown()).optional()
});
const timezoneFieldSchema = fieldConfigSchema.extend({
    type: z.literal('timezone'),
    validateTimezone: z.boolean().prefault(true),
    suggestedValues: z.array(z.string()).optional(),
    restrictToSuggestedValues: z.boolean().prefault(false)
});
const jsonFieldSchema = fieldConfigSchema.extend({
    type: z.literal('json'),
    validateJson: z.boolean().prefault(true)
});
const stringArrayFieldSchema = fieldConfigSchema.extend({
    type: z.literal('string[]')
});
const numberArrayFieldSchema = fieldConfigSchema.extend({
    type: z.literal('number[]')
});
z.discriminatedUnion('type', [
    stringFieldSchema,
    numberFieldSchema,
    booleanFieldSchema,
    dateFieldSchema,
    timezoneFieldSchema,
    jsonFieldSchema,
    stringArrayFieldSchema,
    numberArrayFieldSchema
]);
Object.freeze({
    NOT_FOUND: 'Resource not found',
    BAD_REQUEST: 'Bad request',
    CONFLICT: 'Conflict with current state',
    MISSING_REQUIRED_PARAMETER: 'Missing required parameter',
    UNAUTHORIZED: 'Unauthorized',
    FORBIDDEN: 'Forbidden',
    INTERNAL_SERVER_ERROR: 'Internal server error',
    INITIALIZATION_FAILED: 'Initialization failed',
    DATABASE_CONNECTION_ERROR: 'Database connection error',
    DATABASE_QUERY_ERROR: 'Database query error',
    INVALID_CONFIGURATION: 'Invalid configuration',
    REQUEST_HANDLER_ERROR: 'Request handler error',
    INVALID_REQUEST: 'Invalid request',
    UNKNOWN_ERROR: 'Unknown error',
    NETWORK_ERROR: 'Network error',
    PLUGIN_INITIALIZATION_FAILED: 'Plugin initialization failed',
    API_RETRIEVAL_ERROR: 'API retrieval error',
    VALIDATION_ERROR: 'Validation error',
    UNEXPECTED: 'Unexpected error'
});
Object.freeze({
    VALIDATION: 'validation',
    AUTHORIZATION: 'authorization',
    STORAGE: 'storage',
    NETWORK: 'network',
    PLUGIN: 'plugin',
    CONFIGURATION: 'configuration',
    UNEXPECTED: 'unexpected'
});
trace.getTracer('@doubletie/results');
const schema_auditLogSchema = z.object({
    id: z.string(),
    entityType: z.string(),
    entityId: z.string(),
    actionType: z.string(),
    subjectId: z.string().optional(),
    ipAddress: z.string().optional(),
    userAgent: z.string().optional(),
    changes: z.record(z.string(), z.unknown()).optional(),
    metadata: z.record(z.string(), z.unknown()).optional(),
    createdAt: z.date().prefault(()=>new Date())
});
function getAuditLogTable(options, auditLogFields) {
    const auditLogConfig = options.tables?.auditLog;
    const subjectConfig = options.tables?.subject;
    return {
        entityName: auditLogConfig?.entityName || 'auditLog',
        entityPrefix: auditLogConfig?.entityPrefix || 'log',
        schema: schema_auditLogSchema,
        fields: {
            entityType: {
                type: 'string',
                required: true,
                fieldName: auditLogConfig?.fields?.entityType || 'entityType'
            },
            entityId: {
                type: 'string',
                required: true,
                fieldName: auditLogConfig?.fields?.entityId || 'entityId'
            },
            actionType: {
                type: 'string',
                required: true,
                fieldName: auditLogConfig?.fields?.actionType || 'actionType'
            },
            subjectId: {
                type: 'string',
                required: false,
                fieldName: auditLogConfig?.fields?.subjectId || 'subjectId',
                references: {
                    model: subjectConfig?.entityName || 'subject',
                    field: 'id'
                }
            },
            ipAddress: {
                type: 'string',
                required: false,
                fieldName: auditLogConfig?.fields?.ipAddress || 'ipAddress'
            },
            userAgent: {
                type: 'string',
                required: false,
                fieldName: auditLogConfig?.fields?.userAgent || 'userAgent'
            },
            changes: {
                type: 'json',
                required: false,
                fieldName: auditLogConfig?.fields?.changes || 'changes'
            },
            metadata: {
                type: 'json',
                required: false,
                fieldName: auditLogConfig?.fields?.metadata || 'metadata'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: auditLogConfig?.fields?.createdAt || 'createdAt'
            },
            eventTimezone: {
                type: 'timezone',
                required: true,
                defaultValue: COMMON_TIMEZONES.UTC,
                fieldName: auditLogConfig?.fields?.eventTimezone || 'eventTimezone'
            },
            ...auditLogFields || {},
            ...auditLogConfig?.additionalFields || {}
        },
        indexes: [
            {
                name: 'entity_index',
                fields: [
                    'entityType',
                    'entityId'
                ]
            },
            {
                name: 'action_type_index',
                fields: [
                    'actionType'
                ]
            },
            {
                name: 'subject_id_index',
                fields: [
                    'subjectId'
                ]
            },
            {
                name: 'created_at_index',
                fields: [
                    'createdAt'
                ]
            }
        ],
        order: 5
    };
}
const schema_consentSchema = z.object({
    id: z.string(),
    subjectId: z.string(),
    domainId: z.string(),
    purposeIds: z.array(z.string()),
    metadata: z.record(z.string(), z.unknown()).nullable().optional(),
    policyId: z.string().optional(),
    ipAddress: z.string().nullable().optional(),
    userAgent: z.string().nullable().optional(),
    status: z["enum"]([
        'active',
        'withdrawn',
        'expired'
    ]).prefault('active'),
    withdrawalReason: z.string().nullable().optional(),
    givenAt: z.date().prefault(()=>new Date()),
    validUntil: z.date().nullable().optional(),
    isActive: z.boolean().prefault(true)
});
function getConsentTable(options, consentFields) {
    const consentConfig = options.tables?.consent;
    const subjectConfig = options.tables?.subject;
    const domainConfig = options.tables?.domain;
    const policyConfig = options.tables?.consentPolicy;
    return {
        entityName: consentConfig?.entityName || 'consent',
        entityPrefix: consentConfig?.entityPrefix || 'cns',
        schema: schema_consentSchema,
        fields: {
            subjectId: {
                type: 'string',
                required: true,
                fieldName: consentConfig?.fields?.subjectId || 'subjectId',
                references: {
                    model: subjectConfig?.entityName || 'subject',
                    field: 'id'
                }
            },
            domainId: {
                type: 'string',
                required: true,
                fieldName: consentConfig?.fields?.domainId || 'domainId',
                references: {
                    model: domainConfig?.entityName || 'domain',
                    field: 'id'
                }
            },
            purposeIds: {
                type: 'json',
                required: false,
                fieldName: consentConfig?.fields?.purposeIds || 'purposeIds'
            },
            metadata: {
                type: 'json',
                required: false,
                fieldName: consentConfig?.fields?.metadata || 'metadata'
            },
            policyId: {
                type: 'string',
                required: false,
                fieldName: consentConfig?.fields?.policyId || 'policyId',
                references: {
                    model: policyConfig?.entityName || 'consentPolicy',
                    field: 'id'
                }
            },
            ipAddress: {
                type: 'string',
                required: false,
                fieldName: consentConfig?.fields?.ipAddress || 'ipAddress'
            },
            userAgent: {
                type: 'string',
                required: false,
                fieldName: consentConfig?.fields?.userAgent || 'userAgent'
            },
            status: {
                type: 'string',
                defaultValue: ()=>'active',
                required: true,
                fieldName: consentConfig?.fields?.status || 'status'
            },
            withdrawalReason: {
                type: 'string',
                required: false,
                fieldName: consentConfig?.fields?.withdrawalReason || 'withdrawalReason'
            },
            givenAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: consentConfig?.fields?.givenAt || 'givenAt'
            },
            validUntil: {
                type: 'date',
                required: false,
                fieldName: consentConfig?.fields?.validUntil || 'validUntil',
                transform: {
                    input: (val, data)=>{
                        if (val) return val;
                        const expiresIn = consentConfig?.expiresIn || 31536000;
                        const givenAt = data.givenAt instanceof Date ? data.givenAt : new Date();
                        if (expiresIn > 0) {
                            const validUntil = new Date(givenAt);
                            validUntil.setSeconds(validUntil.getSeconds() + expiresIn);
                            return validUntil;
                        }
                    }
                }
            },
            isActive: {
                type: 'boolean',
                defaultValue: true,
                required: true,
                fieldName: consentConfig?.fields?.isActive || 'isActive'
            },
            ...consentFields || {},
            ...consentConfig?.additionalFields || {}
        },
        order: 3
    };
}
const PolicyTypeSchema = z["enum"]([
    'cookie_banner',
    'privacy_policy',
    'dpa',
    'terms_and_conditions',
    'marketing_communications',
    'age_verification',
    'other'
]);
const schema_consentPolicySchema = z.object({
    id: z.string(),
    version: z.string(),
    type: PolicyTypeSchema,
    name: z.string(),
    effectiveDate: z.date(),
    expirationDate: z.date().nullable().optional(),
    content: z.string(),
    contentHash: z.string(),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
function getConsentPolicyTable(options, policyFields) {
    const consentPolicyConfig = options.tables?.consentPolicy;
    return {
        entityName: consentPolicyConfig?.entityName || 'consentPolicy',
        entityPrefix: consentPolicyConfig?.entityPrefix || 'pol',
        schema: schema_consentPolicySchema,
        fields: {
            version: {
                type: 'string',
                required: true,
                fieldName: consentPolicyConfig?.fields?.version || 'version'
            },
            type: {
                type: 'string',
                required: true,
                fieldName: consentPolicyConfig?.fields?.type || 'type'
            },
            name: {
                type: 'string',
                required: true,
                fieldName: consentPolicyConfig?.fields?.name || 'name'
            },
            effectiveDate: {
                type: 'date',
                required: true,
                fieldName: consentPolicyConfig?.fields?.effectiveDate || 'effectiveDate'
            },
            expirationDate: {
                type: 'date',
                required: false,
                fieldName: consentPolicyConfig?.fields?.expirationDate || 'expirationDate'
            },
            content: {
                type: 'string',
                required: true,
                fieldName: consentPolicyConfig?.fields?.content || 'content'
            },
            contentHash: {
                type: 'string',
                required: true,
                fieldName: consentPolicyConfig?.fields?.contentHash || 'contentHash'
            },
            isActive: {
                type: 'boolean',
                defaultValue: true,
                required: true,
                fieldName: consentPolicyConfig?.fields?.isActive || 'isActive'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: consentPolicyConfig?.fields?.createdAt || 'createdAt'
            },
            ...policyFields || {},
            ...consentPolicyConfig?.additionalFields || {}
        },
        order: 2
    };
}
const schema_purposeSchema = z.object({
    id: z.string(),
    code: z.string(),
    name: z.string(),
    description: z.string(),
    isEssential: z.boolean().prefault(false),
    dataCategory: z.string().nullish(),
    legalBasis: z.string().nullish(),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
function getPurposeTable(options, purposeFields) {
    const purposeConfig = options.tables?.consentPurpose;
    return {
        entityName: purposeConfig?.entityName || 'consentPurpose',
        entityPrefix: purposeConfig?.entityPrefix || 'pur',
        schema: schema_purposeSchema,
        fields: {
            code: {
                type: 'string',
                required: true,
                fieldName: purposeConfig?.fields?.code || 'code'
            },
            name: {
                type: 'string',
                required: true,
                fieldName: purposeConfig?.fields?.name || 'name'
            },
            description: {
                type: 'string',
                required: true,
                fieldName: purposeConfig?.fields?.description || "description"
            },
            isEssential: {
                type: 'boolean',
                defaultValue: ()=>false,
                required: true,
                fieldName: purposeConfig?.fields?.isEssential || 'isEssential'
            },
            dataCategory: {
                type: 'string',
                required: false,
                fieldName: purposeConfig?.fields?.dataCategory || 'dataCategory'
            },
            legalBasis: {
                type: 'string',
                required: false,
                fieldName: purposeConfig?.fields?.legalBasis || 'legalBasis'
            },
            isActive: {
                type: 'boolean',
                defaultValue: true,
                required: true,
                fieldName: purposeConfig?.fields?.isActive || 'isActive'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: purposeConfig?.fields?.createdAt || 'createdAt'
            },
            updatedAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: purposeConfig?.fields?.updatedAt || 'updatedAt'
            },
            ...purposeFields || {},
            ...purposeConfig?.additionalFields || {}
        },
        order: 1
    };
}
const schema_consentRecordSchema = z.object({
    id: z.string(),
    subjectId: z.string(),
    consentId: z.string().optional(),
    actionType: z.string(),
    details: z.record(z.string(), z.unknown()).optional(),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
function getConsentRecordTable(options, recordFields) {
    const recordConfig = options.tables?.record;
    const subjectConfig = options.tables?.subject;
    const consentConfig = options.tables?.consent;
    return {
        entityName: recordConfig?.entityName || 'consentRecord',
        entityPrefix: recordConfig?.entityPrefix || 'rec',
        schema: schema_consentRecordSchema,
        fields: {
            subjectId: {
                type: 'string',
                required: true,
                fieldName: recordConfig?.fields?.subjectId || 'subjectId',
                references: {
                    model: subjectConfig?.entityName || 'subject',
                    field: 'id'
                }
            },
            consentId: {
                type: 'string',
                required: false,
                fieldName: recordConfig?.fields?.consentId || 'consentId',
                references: {
                    model: consentConfig?.entityName || 'consent',
                    field: 'id'
                }
            },
            actionType: {
                type: 'string',
                required: true,
                fieldName: recordConfig?.fields?.actionType || 'actionType'
            },
            details: {
                type: 'json',
                required: false,
                fieldName: recordConfig?.fields?.details || 'details'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: recordConfig?.fields?.createdAt || 'createdAt'
            },
            ...recordFields || {},
            ...recordConfig?.additionalFields || {}
        },
        order: 4
    };
}
const schema_domainSchema = z.object({
    id: z.string(),
    name: z.string().min(1),
    description: z.string().optional(),
    allowedOrigins: z.array(z.string()).optional().prefault([]),
    isVerified: z.boolean().prefault(true),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
function getDomainTable(options, domainFields) {
    const domainConfig = options.tables?.domain;
    return {
        entityName: domainConfig?.entityName || 'domain',
        entityPrefix: domainConfig?.entityPrefix || 'dom',
        schema: schema_domainSchema,
        fields: {
            name: {
                type: 'string',
                required: true,
                unique: true,
                fieldName: domainConfig?.fields?.name || 'name'
            },
            description: {
                type: 'string',
                required: false,
                fieldName: domainConfig?.fields?.description || "description"
            },
            allowedOrigins: {
                type: 'json',
                defaultValue: ()=>[],
                required: false,
                fieldName: domainConfig?.fields?.allowedOrigins || 'allowedOrigins'
            },
            isVerified: {
                type: 'boolean',
                defaultValue: true,
                required: true,
                fieldName: domainConfig?.fields?.isVerified || 'isVerified'
            },
            isActive: {
                type: 'boolean',
                defaultValue: true,
                required: true,
                fieldName: domainConfig?.fields?.isActive || 'isActive'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: domainConfig?.fields?.createdAt || 'createdAt'
            },
            updatedAt: {
                type: 'date',
                required: false,
                fieldName: domainConfig?.fields?.updatedAt || 'updatedAt'
            },
            ...domainFields || {},
            ...domainConfig?.additionalFields || {}
        },
        order: 1
    };
}
const schema_subjectSchema = z.object({
    id: z.string(),
    isIdentified: z.boolean().prefault(false),
    externalId: z.string().nullable().optional(),
    identityProvider: z.string().optional(),
    lastIpAddress: z.string().optional(),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
function getSubjectTable(options, subjectFields) {
    const subjectConfig = options.tables?.subject;
    return {
        entityName: subjectConfig?.entityName || 'subject',
        entityPrefix: subjectConfig?.entityPrefix || 'sub',
        schema: schema_subjectSchema,
        fields: {
            isIdentified: {
                type: 'boolean',
                defaultValue: ()=>false,
                required: true,
                fieldName: subjectConfig?.fields?.isIdentified || 'isIdentified'
            },
            externalId: {
                type: 'string',
                required: false,
                fieldName: subjectConfig?.fields?.externalId || 'externalId'
            },
            identityProvider: {
                type: 'string',
                required: false,
                fieldName: subjectConfig?.fields?.identityProvider || 'identityProvider'
            },
            lastIpAddress: {
                type: 'string',
                required: false,
                fieldName: subjectConfig?.fields?.lastIpAddress || 'lastIpAddress'
            },
            createdAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: subjectConfig?.fields?.createdAt || 'createdAt'
            },
            updatedAt: {
                type: 'date',
                defaultValue: ()=>new Date(),
                required: true,
                fieldName: subjectConfig?.fields?.updatedAt || 'updatedAt'
            },
            subjectTimezone: {
                type: 'timezone',
                required: false,
                defaultValue: COMMON_TIMEZONES.UTC,
                fieldName: subjectConfig?.fields?.subjectTimezone || 'subjectTimezone'
            },
            ...subjectFields || {},
            ...subjectConfig?.additionalFields || {}
        },
        order: 1
    };
}
const definition_getConsentTables = (options)=>{
    const pluginSchema = options.plugins?.reduce((acc, plugin)=>{
        const schema = plugin.schema;
        if (!schema) return acc;
        for (const [key, value] of Object.entries(schema))acc[key] = {
            fields: {
                ...acc[key]?.fields,
                ...value.fields
            },
            entityName: key
        };
        return acc;
    }, {});
    const { subject, consentPurpose, consentPolicy, domain, geoLocation, consent, consentPurposeJunction, record, consentGeoLocation, consentWithdrawal, auditLog, ...pluginTables } = pluginSchema || {};
    return {
        subject: getSubjectTable(options, subject?.fields),
        consentPurpose: getPurposeTable(options, consentPurpose?.fields),
        consentPolicy: getConsentPolicyTable(options, consentPolicy?.fields),
        domain: getDomainTable(options, domain?.fields),
        consent: getConsentTable(options, consent?.fields),
        consentRecord: getConsentRecordTable(options, record?.fields),
        auditLog: getAuditLogTable(options, auditLog?.fields),
        ...pluginTables
    };
};
function utils_applyDefaultValue(inputValue, field, operation) {
    if ('update' === operation) return inputValue;
    if (null == inputValue && field.defaultValue) {
        if ('function' == typeof field.defaultValue) return field.defaultValue();
        return field.defaultValue;
    }
    return inputValue;
}
const kysely_adapter_createEntityTransformer = (db, options, config)=>{
    const schema = definition_getConsentTables(options);
    function getField(model, field) {
        if ('id' === field) return field;
        const modelFields = schema[model]?.fields;
        const f = modelFields ? modelFields[field] : void 0;
        if (!f) console.log('Field not found', model, field);
        return f?.fieldName || field;
    }
    function transformValueToDB(value, model, field) {
        if ('id' === field) return value;
        const { type = 'sqlite' } = config || {};
        const modelFields = schema[model]?.fields;
        const f = modelFields ? modelFields[field] : void 0;
        if (f?.type === 'boolean' && ('sqlite' === type || 'mssql' === type) && null != value) return value ? 1 : 0;
        if (f?.type === 'date' && value && value instanceof Date) return 'sqlite' === type ? value.toISOString() : value;
        if (f?.type === 'json' && null != value) {
            if ('postgres' === type || 'mysql' === type) return value;
            return superjson.stringify(value);
        }
        return value;
    }
    function transformValueFromDB(value, model, field) {
        const { type = 'sqlite' } = config || {};
        const modelFields = schema[model]?.fields;
        const f = modelFields ? modelFields[field] : void 0;
        if (f?.type === 'boolean' && ('sqlite' === type || 'mssql' === type) && null !== value) return 1 === value;
        if (f?.type === 'date' && value) return new Date(value);
        if (f?.type === 'json' && null != value) {
            if (('postgres' === type || 'mysql' === type) && 'object' == typeof value) return value;
            if ('string' == typeof value) try {
                return superjson.parse(value);
            } catch  {
                try {
                    return JSON.parse(value);
                } catch  {}
            }
        }
        return value;
    }
    function getEntityName(model) {
        return schema[model].entityName;
    }
    return {
        transformInput (data, model, action) {
            const transformedData = {};
            if ('create' === action) {
                const advancedOptions = options.advanced || {};
                transformedData.id = data.id || (advancedOptions.generateId ? advancedOptions.generateId({
                    model
                }) : id_generator_generateId(schema[model].entityPrefix));
            }
            const fields = schema[model].fields;
            for(const field in fields)if (Object.hasOwn(fields, field)) {
                const value = data[field];
                const fieldInfo = fields[field];
                const fieldName = fieldInfo?.fieldName || field;
                if (fieldInfo) transformedData[fieldName] = utils_applyDefaultValue(transformValueToDB(value, model, field), fieldInfo, action);
            }
            return transformedData;
        },
        transformOutput (data, model, select = []) {
            if (!data) return null;
            const transformedData = {};
            if (data.id && (0 === select.length || select.includes('id'))) transformedData.id = data.id;
            const tableSchema = schema[model]?.fields;
            for(const key in tableSchema){
                if (select.length && !select.includes(key)) continue;
                const field = tableSchema[key];
                if (field) transformedData[key] = transformValueFromDB(data[field.fieldName || key], model, key);
            }
            return transformedData;
        },
        convertWhereClause (model, whereConditions) {
            if (!whereConditions || 0 === whereConditions.length) return {
                and: null,
                or: null
            };
            const conditions = {
                and: [],
                or: []
            };
            for (const condition of whereConditions){
                let { field: _field, value, operator = '=', connector = 'AND' } = condition;
                const fieldString = getField(model, _field);
                value = transformValueToDB(value, model, _field);
                const expr = (eb)=>{
                    const dbField = fieldString;
                    if ('in' === operator.toLowerCase()) return eb(dbField, 'in', Array.isArray(value) ? value : [
                        value
                    ]);
                    if ('contains' === operator) return eb(dbField, 'like', `%${value}%`);
                    if ('starts_with' === operator) return eb(dbField, 'like', `${value}%`);
                    if ('ends_with' === operator) return eb(dbField, 'like', `%${value}`);
                    if ('ilike' === operator) {
                        const lowerField = eb.fn('lower', [
                            dbField
                        ]);
                        const lowerValue = eb.fn('lower', [
                            eb.val(value?.toString())
                        ]);
                        return eb(lowerField, 'like', lowerValue);
                    }
                    if ('eq' === operator) return eb(dbField, '=', value);
                    if ('ne' === operator) return eb(dbField, '<>', value);
                    if ('gt' === operator) return eb(dbField, '>', value);
                    if ('gte' === operator) return eb(dbField, '>=', value);
                    if ('lt' === operator) return eb(dbField, '<', value);
                    if ('lte' === operator) return eb(dbField, '<=', value);
                    return eb(dbField, operator, value);
                };
                if ('OR' === connector) conditions.or.push(expr);
                else conditions.and.push(expr);
            }
            return {
                and: conditions.and.length ? conditions.and : null,
                or: conditions.or.length ? conditions.or : null
            };
        },
        async withReturning (values, builder, model, where) {
            let res = null;
            if (config?.type === 'mysql') {
                await builder.execute();
                const whereCondition = where[0];
                const field = values.id ? 'id' : whereCondition?.field ?? 'id';
                const value = values[field] ?? whereCondition?.value;
                const fieldString = getField(model, field);
                res = await db.selectFrom(getEntityName(model)).selectAll().where((eb)=>eb(fieldString, '=', value)).executeTakeFirst();
                return res;
            }
            if (config?.type === 'mssql') {
                res = await builder.outputAll('inserted').executeTakeFirst();
                return res;
            }
            res = await builder.returningAll().executeTakeFirst();
            return res;
        },
        getEntityName,
        getField
    };
};
const kysely_adapter_kyselyAdapter = (db, config)=>(opts)=>{
        const { transformInput, withReturning, transformOutput, convertWhereClause, getEntityName, getField } = kysely_adapter_createEntityTransformer(db, opts, config);
        return {
            id: 'kysely',
            async create (data) {
                const { model, data: values, select } = data;
                const transformed = transformInput(values, model, 'create');
                const tableName = getEntityName(model);
                const builder = db.insertInto(tableName).values(transformed);
                const result = await withReturning(transformed, builder, model, []);
                const output = transformOutput(result, model, select);
                return output;
            },
            async findOne (data) {
                const { model, where, select } = data;
                const whereArray = Array.isArray(where) ? where : [
                    where
                ];
                const { and, or } = convertWhereClause(model, whereArray);
                const tableName = getEntityName(model);
                let query = db.selectFrom(tableName).selectAll();
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                const res = await query.executeTakeFirst();
                if (!res) return null;
                return transformOutput(res, model, select);
            },
            async findMany (data) {
                const { model, where, limit, offset, sortBy } = data;
                const whereArray = where ? Array.isArray(where) ? where : [
                    where
                ] : void 0;
                const { and, or } = convertWhereClause(model, whereArray);
                const tableName = getEntityName(model);
                let query = db.selectFrom(tableName);
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                if (config?.type === 'mssql') {
                    if (!offset) query = query.top(limit || 100);
                } else query = query.limit(limit || 100);
                if (sortBy) {
                    const sortFieldString = getField(model, sortBy.field);
                    query = query.orderBy(sortFieldString, sortBy.direction);
                }
                if (offset) if (config?.type === 'mssql') {
                    if (!sortBy) query = query.orderBy('id');
                    query = query.offset(offset).fetch(limit || 100);
                } else query = query.offset(offset);
                const res = await query.selectAll().execute();
                if (!res) return [];
                return res.map((r)=>transformOutput(r, model));
            },
            async update (data) {
                const { model, where, update: values } = data;
                const whereArray = Array.isArray(where) ? where : [
                    where
                ];
                const { and, or } = convertWhereClause(model, whereArray);
                const transformedData = transformInput(values, model, 'update');
                const tableName = getEntityName(model);
                let query = db.updateTable(tableName).set(transformedData);
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                const result = await withReturning(transformedData, query, model, whereArray);
                return transformOutput(result, model);
            },
            async updateMany (data) {
                const { model, where, update: values } = data;
                const whereArray = Array.isArray(where) ? where : [
                    where
                ];
                const { and, or } = convertWhereClause(model, whereArray);
                const transformedData = transformInput(values, model, 'update');
                const tableName = getEntityName(model);
                let query = db.updateTable(tableName).set(transformedData);
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                await query.execute();
                let selectQuery = db.selectFrom(tableName).selectAll();
                if (and) selectQuery = selectQuery.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) selectQuery = selectQuery.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                const fetchedResults = await selectQuery.execute();
                if (!fetchedResults || 0 === fetchedResults.length) return [];
                return fetchedResults.map((record)=>transformOutput(record, model));
            },
            async count (data) {
                const { model, where } = data;
                const whereArray = where ? Array.isArray(where) ? where : [
                    where
                ] : void 0;
                const { and, or } = convertWhereClause(model, whereArray);
                const tableName = getEntityName(model);
                let query = db.selectFrom(tableName).select((eb)=>eb.fn.count('id').as('count'));
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                const res = await query.execute();
                const count = res[0]?.count;
                return 'number' == typeof count ? count : 0;
            },
            async delete (data) {
                const { model, where } = data;
                const whereArray = Array.isArray(where) ? where : [
                    where
                ];
                const { and, or } = convertWhereClause(model, whereArray);
                const tableName = getEntityName(model);
                let query = db.deleteFrom(tableName);
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                await query.execute();
            },
            async deleteMany (data) {
                const { model, where } = data;
                const whereArray = Array.isArray(where) ? where : [
                    where
                ];
                const { and, or } = convertWhereClause(model, whereArray);
                const tableName = getEntityName(model);
                let query = db.deleteFrom(tableName);
                if (and) query = query.where((eb)=>{
                    const conditions = and.map((expr)=>expr(eb));
                    return eb.and(conditions);
                });
                if (or) query = query.where((eb)=>{
                    const conditions = or.map((expr)=>expr(eb));
                    return eb.or(conditions);
                });
                const result = await query.execute();
                const count = result.length;
                return count;
            },
            async transaction (data) {
                const { callback } = data;
                const advancedOptions = opts.advanced || {};
                if (advancedOptions.disableTransactions) {
                    const regularAdapter = kysely_adapter_kyselyAdapter(db, config)(opts);
                    return await callback(regularAdapter);
                }
                try {
                    return await db.transaction().execute(async (trx)=>{
                        const transactionAdapter = kysely_adapter_kyselyAdapter(trx, config)(opts);
                        return await callback(transactionAdapter);
                    });
                } catch (error) {
                    if (error instanceof Error && (error.message.includes('transactions are not supported') || error.message.toLowerCase().includes('no transaction support'))) {
                        console.warn("Warning: Database transaction failed. If your database does not support transactions, you can disable this warning by setting opts.advanced.disableTransactions to true.");
                        const regularAdapter = kysely_adapter_kyselyAdapter(db, config)(opts);
                        return await callback(regularAdapter);
                    }
                    throw error;
                }
            },
            options: config
        };
    };
export { dialect_createKyselyAdapter as createKyselyAdapter, kysely_adapter_kyselyAdapter as kyselyAdapter };
