// @c15t/logger/dist/core/logger.js
import { getFormatter } from "../formatting/formatter.js";
import { levels, shouldPublishLog } from "./levels.js";
const createLogger = (options)=>{
    if (options && 'object' == typeof options && levels.every((level)=>'function' == typeof options[level])) return options;
    const loggerOptions = options;
    const enabled = loggerOptions?.disabled !== true;
    const logLevel = loggerOptions?.level ?? 'error';
    const appName = loggerOptions?.appName ?? 'c15t';
    const consoleMethods = {
        error: console.error.bind(console),
        warn: console.warn.bind(console),
        info: console.log.bind(console),
        debug: console.debug.bind(console),
        success: console.log.bind(console)
    };
    const logFunc = (level, message, args = [])=>{
        if (!enabled || !shouldPublishLog(logLevel, level)) return;
        const formatter = getFormatter('default');
        const formattedMessage = formatter(level, message, args, appName);
        if (!loggerOptions || 'function' != typeof loggerOptions.log) {
            const consoleMethod = consoleMethods[level];
            consoleMethod(formattedMessage, ...args);
            return;
        }
        loggerOptions.log('success' === level ? 'info' : level, message, ...args);
    };
    return Object.fromEntries(levels.map((level)=>[
            level,
            (...[message, ...args])=>{
                logFunc(level, message, args);
            }
        ]));
};
const logger = createLogger();
function extendLogger(baseLogger, extensions) {
    return Object.assign({}, baseLogger, extensions);
}
export { createLogger, extendLogger, logger };
