// @c15t/backend/dist/v2/db/schema/index.js
import { fumadb } from "fumadb";
import { column, idColumn, schema, table } from "fumadb/schema";
import { z } from "zod";
const auditLogTable = table('auditLog', {
    id: idColumn('id', 'varchar(255)'),
    entityType: column('entityType', 'string'),
    entityId: column('entityId', 'string'),
    actionType: column('actionType', 'string'),
    subjectId: column('subjectId', 'string').nullable(),
    ipAddress: column('ipAddress', 'string').nullable(),
    userAgent: column('userAgent', 'string').nullable(),
    changes: column('changes', 'json').nullable(),
    metadata: column('metadata', 'json').nullable(),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now'),
    eventTimezone: column('eventTimezone', 'string').defaultTo$(()=>'UTC')
});
const auditLogSchema = z.object({
    id: z.string(),
    entityType: z.string(),
    entityId: z.string(),
    actionType: z.string(),
    subjectId: z.string().optional(),
    ipAddress: z.string().optional(),
    userAgent: z.string().optional(),
    changes: z.record(z.string(), z.unknown()).optional(),
    metadata: z.record(z.string(), z.unknown()).optional(),
    createdAt: z.date().prefault(()=>new Date()),
    eventTimezone: z.string().prefault('UTC')
});
const consentTable = table('consent', {
    id: idColumn('id', 'varchar(255)'),
    subjectId: column('subjectId', 'string'),
    domainId: column('domainId', 'string'),
    policyId: column('policyId', 'string').nullable(),
    purposeIds: column('purposeIds', 'json'),
    metadata: column('metadata', 'json').nullable(),
    ipAddress: column('ipAddress', 'string').nullable(),
    userAgent: column('userAgent', 'string').nullable(),
    status: column('status', 'string').defaultTo$(()=>'active'),
    withdrawalReason: column('withdrawalReason', 'string').nullable(),
    givenAt: column('givenAt', 'timestamp').defaultTo$('now'),
    validUntil: column('validUntil', 'timestamp').nullable(),
    isActive: column('isActive', 'bool').defaultTo$(()=>true)
});
const consentStatusSchema = z["enum"]([
    'active',
    'withdrawn',
    'expired'
]);
const consentSchema = z.object({
    id: z.string(),
    subjectId: z.string(),
    domainId: z.string(),
    purposeIds: z.array(z.string()),
    metadata: z.record(z.string(), z.unknown()).nullish(),
    policyId: z.string().optional(),
    ipAddress: z.string().nullish(),
    userAgent: z.string().nullish(),
    status: consentStatusSchema.prefault('active'),
    withdrawalReason: z.string().nullish(),
    givenAt: z.date().prefault(()=>new Date()),
    validUntil: z.date().nullish(),
    isActive: z.boolean().prefault(true)
});
const consentPolicyTable = table('consentPolicy', {
    id: idColumn('id', 'varchar(255)'),
    version: column('version', 'string'),
    type: column('type', 'string'),
    name: column('name', 'string'),
    effectiveDate: column('effectiveDate', 'timestamp'),
    expirationDate: column('expirationDate', 'timestamp').nullable(),
    content: column('content', 'string'),
    contentHash: column('contentHash', 'string'),
    isActive: column('isActive', 'bool').defaultTo$(()=>true),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now')
});
const PolicyTypeSchema = z["enum"]([
    'cookie_banner',
    'privacy_policy',
    'dpa',
    'terms_and_conditions',
    'marketing_communications',
    'age_verification',
    'other'
]);
const consentPolicySchema = z.object({
    id: z.string(),
    version: z.string(),
    type: PolicyTypeSchema,
    name: z.string(),
    effectiveDate: z.date(),
    expirationDate: z.date().nullish(),
    content: z.string(),
    contentHash: z.string(),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date())
});
const consentPurposeTable = table('consentPurpose', {
    id: idColumn('id', 'varchar(255)'),
    code: column('code', 'string'),
    name: column('name', 'string'),
    description: column("description", 'string'),
    isEssential: column('isEssential', 'bool'),
    dataCategory: column('dataCategory', 'string').nullable(),
    legalBasis: column('legalBasis', 'string').nullable(),
    isActive: column('isActive', 'bool').defaultTo$(()=>true),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now'),
    updatedAt: column('updatedAt', 'timestamp').defaultTo$('now')
});
const consentPurposeSchema = z.object({
    id: z.string(),
    code: z.string(),
    name: z.string(),
    description: z.string(),
    isEssential: z.boolean(),
    dataCategory: z.string().nullish(),
    legalBasis: z.string().nullish(),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
const consentRecordTable = table('consentRecord', {
    id: idColumn('id', 'varchar(255)'),
    subjectId: column('subjectId', 'string'),
    consentId: column('consentId', 'string').nullable(),
    actionType: column('actionType', 'string'),
    details: column('details', 'json').nullable(),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now')
});
const consentRecordSchema = z.object({
    id: z.string(),
    subjectId: z.string(),
    consentId: z.string().nullish(),
    actionType: z.string(),
    details: z.record(z.string(), z.unknown()).nullish(),
    createdAt: z.date().prefault(()=>new Date())
});
const domainTable = table('domain', {
    id: idColumn('id', 'varchar(255)'),
    name: column('name', 'string').unique(),
    description: column("description", 'string').nullable(),
    allowedOrigins: column('allowedOrigins', 'json').nullable(),
    isVerified: column('isVerified', 'bool').defaultTo$(()=>true),
    isActive: column('isActive', 'bool').defaultTo$(()=>true),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now'),
    updatedAt: column('updatedAt', 'timestamp').defaultTo$('now')
});
const domainSchema = z.object({
    id: z.string(),
    name: z.string(),
    description: z.string().nullish(),
    allowedOrigins: z.array(z.string()).nullish(),
    isVerified: z.boolean().prefault(true),
    isActive: z.boolean().prefault(true),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
const subjectTable = table('subject', {
    id: idColumn('id', 'varchar(255)'),
    isIdentified: column('isIdentified', 'bool').defaultTo$(()=>false),
    externalId: column('externalId', 'string').nullable(),
    identityProvider: column('identityProvider', 'string').nullable(),
    lastIpAddress: column('lastIpAddress', 'string').nullable(),
    subjectTimezone: column('subjectTimezone', 'string').nullable(),
    createdAt: column('createdAt', 'timestamp').defaultTo$('now'),
    updatedAt: column('updatedAt', 'timestamp').defaultTo$('now')
});
const subjectSchema = z.object({
    id: z.string(),
    isIdentified: z.boolean().prefault(false),
    externalId: z.string().nullish(),
    identityProvider: z.string().nullish(),
    lastIpAddress: z.string().optional(),
    subjectTimezone: z.string().nullish(),
    createdAt: z.date().prefault(()=>new Date()),
    updatedAt: z.date().prefault(()=>new Date())
});
const v1 = schema({
    version: '1.0.0',
    tables: {
        subject: subjectTable,
        domain: domainTable,
        consentPolicy: consentPolicyTable,
        consentPurpose: consentPurposeTable,
        consent: consentTable,
        auditLog: auditLogTable,
        consentRecord: consentRecordTable
    },
    relations: {
        subject: ({ many })=>({
                consents: many('consent'),
                consentRecords: many('consentRecord'),
                auditLogs: many('auditLog')
            }),
        domain: ({ many })=>({
                consents: many('consent')
            }),
        consentPolicy: ({ many })=>({
                consents: many('consent')
            }),
        consentPurpose: ()=>({}),
        consent: ({ one, many })=>({
                subject: one('subject', [
                    'subjectId',
                    'id'
                ]).foreignKey(),
                domain: one('domain', [
                    'domainId',
                    'id'
                ]).foreignKey(),
                policy: one('consentPolicy', [
                    'policyId',
                    'id'
                ]).foreignKey(),
                consentRecords: many('consentRecord')
            }),
        consentRecord: ({ one })=>({
                subject: one('subject', [
                    'subjectId',
                    'id'
                ]).foreignKey(),
                consent: one('consent', [
                    'consentId',
                    'id'
                ]).foreignKey()
            }),
        auditLog: ({ one })=>({
                subject: one('subject', [
                    'subjectId',
                    'id'
                ]).foreignKey()
            })
    }
});
const DB = fumadb({
    namespace: 'c15t',
    schemas: [
        v1
    ]
});
export { DB, PolicyTypeSchema, auditLogSchema, auditLogTable, consentPolicySchema, consentPolicyTable, consentPurposeSchema, consentPurposeTable, consentRecordSchema, consentRecordTable, consentSchema, consentStatusSchema, consentTable, domainSchema, domainTable, subjectSchema, subjectTable, v1 };
